Update the GASNet package

bonachea opened the pull request:
The previous `gasnet` spack package was not vetted/approved by the
GASNet library maintainers. This one is.

Notably adds build-time testing and smoke-testing.

bonachea committed:
gasnet: Rework the GASNet package

The previous `gasnet` spack package was not vetted/approved by the
GASNet library maintainers. This one is.

Notably adds build-time testing and smoke-testing.

bonachea committed:
upcxx: Add variant to override GASNet-EX version

eugeneswalker commented:
I used these modifications to build `gasnet` inside our E4S Spack environment for both x86_64 and ppc64le for the following:
* Ubuntu 18.04 GCC 7.5.0
* Ubuntu 20.04 GCC 9.3.0
* RHEL 7 GCC 9.3.0
* RHEL 8 GCC 8.3.1

eugeneswalker commented:
This looks good to me. I'd like to give @tldahlgren a chance to weigh on the Spack test aspect of this before merging.

Thank you for the updates!

tldahlgren commented on var/spack/repos/builtin/packages/gasnet/package.py:
```diff
@@ -6,98 +6,133 @@
+    version('2020.10.0', sha256='ed17baf7fce90499b539857ee37b3eea961aa475cffbde77e4c607a34ece06a0')
+    version('2020.3.0',  sha256='019eb2d2284856e6fabe6c8c0061c874f10e95fa0265245f227fd3497f1bb274')
+    version('2019.9.0',  sha256='117f5fdb16e53d0fa8a47a1e28cccab1d8020ed4f6e50163d985dc90226aaa2c')
+    # Do NOT add older versions here.
```
We discourage removing older versions from the package. 

People will still be able to use Spack to install any available versions but, without the `version` directive, they'll get ``There is no checksum on file to fetch gasnet@<old-version> safely. Fetch anyway?`` prompts.

Is that behavior desirable?

bonachea replied:
We (the library maintainers) are not supporting versions more than 2 years old. Part of this is due to changes in configure over time that would complicate package maintenance, but more importantly we want to strongly discourage use of outdated software.

I agree that nothing prevents people from fetching any ancient version they want, but we don't want to implicitly endorse/encourage that behavior, so I'm fine with a warning prompt if someone explicitly requests that.

tldahlgren commented on var/spack/repos/builtin/packages/gasnet/package.py:
```diff
@@ -6,98 +6,133 @@
+            env['GASNET_SSH_SERVERS'] = "localhost " * 4
+
+        if len(self.conduits()) > 0:
+            self.run_test(join_path(prefix, 'tests', 'testtools'),
```
Same issue mentioned elsewhere:

```suggestion
            self.run_test(join_path(self.prefix.tests, 'testtools'),
```

bonachea replied:
Fixed as suggested in 83f9bf4

tldahlgren commented on var/spack/repos/builtin/packages/gasnet/package.py:
```diff
@@ -6,98 +6,133 @@
+        install_tree('.', prefix + "/src")
+
+        # Library build is provided for unit-testing purposes only (see notice above)
+        if len(self.conduits()) > 0:
```
Nothing gets installed -- except the source files from above -- if none of the conduit options are enabled.  Is that the desired behavior?

bonachea replied:
Correct, that's intentional. 

This is exactly the capability desired by production GASNet client packages (such as UPC++), that can use this feature to configure and build their embedded GASNet-EX libraries from these source files using a Spack-directed version. For example, this PR enables one to easily build UPC++ with any selected GASNet-EX version, via a command like:

`spack install upcxx+gasnet^gasnet@develop`

I realize this might not be exactly common Spack practice, but it enables the intended capability which is desirable for debugging/support scenarios (in our admittedly unusual build paradigm).

becker33 commented on var/spack/repos/builtin/packages/gasnet/package.py:
```diff
@@ -6,98 +6,133 @@
-
-    conflicts('+aligned-segments', when='+pshm')
+    version('develop', branch='develop')
+    version('main',    branch='stable')
```
Is `stable` newer or older than `master`? If it's older, then `trunk` may be a better match for what you want, because `develop > main >  master > head > trunk > 99999 > 0 > zzzzz > a`  is how we sort version tuples.

bonachea replied:
Yes this was intentional - our stable branch is generally newer than master (although there are times, like now, when they are textually equal).

becker33 commented on var/spack/repos/builtin/packages/gasnet/package.py:
```diff
@@ -6,98 +6,133 @@
-        else:
-            args.append('--disable-aligned-segments')
+    # The optional network backends:
+    variant('smp', default=True, description="SMP conduit for single-node operation")
+    variant('mpi', default=False, description="Low-performance/portable MPI conduit")
+    variant('ibv', default=False, description="Native InfiniBand verbs conduit")
+    variant('udp', default=False, description="Portable UDP conduit, for Ethernet networks")
```
This should be a single variant
```
variant('conduits', values=any_combination_of('smp', 'mpi', 'ibv', 'udp').with_default('smp'), 
             description=...)
```

bonachea replied:
This would be a major departure from the variants in the package this is replacing, but I'm fine implementing that if this is the consensus view?

bonachea replied:
IIUC One significant downside of this suggestion is it would narrow/limit the amount of user documentation the package can conveniently provide for each value in this most critical setting.

Current `spack info gasnet`:
```
Variants:
    Name [Default]    Allowed values    Description
    ==============    ==============    ===========================================

    debug [off]       on, off           Enable library debugging mode
    ibv [off]         on, off           Native InfiniBand verbs conduit
    mpi [off]         on, off           Low-performance/portable MPI conduit
    smp [on]          on, off           SMP conduit for single-node operation
    udp [off]         on, off           Portable UDP conduit, for Ethernet networks
```


Proposed `spack info gasnet`:

```
Variants:
    Name [Default]    Allowed values          Description
    ==============    ====================    ===========================================================================

    conduits [smp]    none, udp, smp, mpi,    The hardware-dependent network backends to enable. For detailed
                      ibv                     recommendations, consult https://gasnet.lbl.gov
    debug [off]       on, off                 Enable library debugging mode
```

A novice user has some chance of recognizing the correct/relevant network hardware backend from the first info output, but the second version sacrifices that information and might lead to a user making a less appropriate choice (that could either fail to build, fail at runtime, or fail to best leverage the available network hardware). 

Is there any way to provide better/formatted info documentation for each value in a `values=any_combination_of`?

PHHargrove replied:
As @bonachea's co-developer on GASNet, I too would prefer the richer `info` output afforded by the current variants over the alternative.  When the end user who has InfiniBand hardware gets it wrong, they could easily get correct behavior from udp or mpi but _horrible_ performance.   The current descriptions go a long way toward addressing that sort of thing.

becker33 replied:
There is no limit to the length of the docstring you give for a variant. You can add  information on every conduit option to the variant description if that will be useful for users.

bonachea replied:
becker33 said:
> There is no limit to the length of the docstring you give for a variant

Unfortunately `spack info` appears to squash/strip all formatting whitespace, so there's no clean way to present a bulleted list of value/description pairs.

I've gone ahead and made the requested change in cae743d, although I think the resulting output is less readable:
```
Variants:
    Name [Default]    Allowed values          Description
    ==============    ====================    ==================================

    conduits [smp]    none, mpi, ibv, smp,    The hardware-dependent network
                      udp                     backends to enable. (smp) = SMP
                                              conduit for single-node operation
                                              ; (ibv) = Native InfiniBand verbs
                                              conduit ; (udp) = Portable UDP
                                              conduit, for Ethernet networks ;
                                              (mpi) = Low-performance/portable
                                              MPI conduit ; For detailed
                                              recommendations, consult
                                              https://gasnet.lbl.gov
    debug [off]       on, off                 Enable library debugging mode
```

becker33 reviewed (changes requested):
This was only a partial review with what I noticed as I went.

tldahlgren commented on var/spack/repos/builtin/packages/gasnet/package.py:
```diff
@@ -6,98 +6,133 @@
+        if '+smp' in self.spec:
+            make('-C', 'smp-conduit', 'run-tests')
+        if len(self.conduits()) > 0:
+            self.run_test(join_path(prefix, 'tests', 'testtools'),
+                          expected=['Done.'], status=0,
+                          installed=True, purpose="Running testtools")
```
This method is intended for use with stand-alone/smoke tests. 

While the test will run after installation, there's a fair amount of overhead using it in this context over executing testtools directly *unless* you really want the output and status comparisons, in which case they do provide helpful checks even in this context.

bonachea replied:
The output and status comparisons are the reason I wrote it this way. Otherwise a detected error seems more likely to go unnoticed. The test itself runs for about 20 seconds, so I'm not too worried about a small amount of parsing overhead.

On a related note, is there any way to have `run_test` look for _failure_ strings (eg the word "ERROR") in the output to trigger a test failure? The [documentation](https://spack.readthedocs.io/en/latest/packaging_guide.html#adding-stand-alone-tests) seems to imply this might be possible, but doesn't explain how.

tldahlgren replied:
Did you try to provide it as expected output and it didn't work?  

Does the failure also return a non-zero status value?  If so, then you'll want to provide that as the expected status.  

If the failure could return different non-zero status values across versions -- or 0 for some versions and non-zero for others -- you could provide a list of "acceptable" status values.  We didn't get very explicit in this case in part because we weren't sure how often such a feature would be used.

tldahlgren replied:
Let us know if you want to attempt this check in the PR or in a subsequent one.

bonachea replied:
Hi @tldahlgren - I think you misunderstood my question. 

Our test prints messages (like "Done.") that indicate the test completed running. The test is also capable of printing messages (like "ERROR") that indicate a failure of a sub-test was detected and the test run should be considered a failure. In other words, the test's success condition is something like:  
```
(<test-output> contains <expected-string>) AND NOT (<test-output> contains <error-string>)
```

IIUC, the `run_test` interface handles the first condition, but it does not allow me to specify the latter (ie outputs that trigger test failure)..

I suspect this is a capability currently lacking in `run_test`, so I'm fine deferring this particular detail to a later PR.

tldahlgren commented on var/spack/repos/builtin/packages/gasnet/package.py:
```diff
@@ -6,98 +6,133 @@
+        if '+smp' in self.spec:
+            make('-C', 'smp-conduit', 'run-tests')
+        if len(self.conduits()) > 0:
+            self.run_test(join_path(prefix, 'tests', 'testtools'),
```
While not technically necessary thanks to a bit of behind the scenes Spack magic, I prefer to be more explicit (to avoid potentially developing bad habits :grin:) in places like this, such as:

```suggestion
            self.run_test(join_path(self.prefix.tests, 'testtools'),
```

bonachea replied:
Fixed as suggested in 83f9bf4

tldahlgren commented on var/spack/repos/builtin/packages/gasnet/package.py:
```diff
@@ -6,98 +6,133 @@
+            'udp': [join_path(self.prefix.bin, 'amudprun'), '-spawn', 'L', '-np', ranks]
+        }
+
+        env['GASNET_VERBOSEENV'] = '1'  # include diagnostic info
```
FWIW. I tend to prefer being explicit wrt the environment (`os.environ`) as well but am probably pushing it in this case. :grin:

bonachea replied:
Fixed as suggested in 83f9bf4

bonachea committed:
gasnet: Respond to PR feedback

* Use `self.prefix.tests`
* Use `os.environ` explicitly

bonachea committed:
gasnet: Convert network variants into a multi-valued `conduits` variant

This has the minor advantage of enabling a concise `conduits=none` spec,
but the major drawback that it degrades the `spack info gasnet` output.

becker33 commented on var/spack/repos/builtin/packages/upcxx/package.py:
```diff
@@ -150,8 +159,12 @@ def install(self, spec, prefix):
+            options = ["--prefix=%s" % prefix]
+
+            if '+gasnet' in self.spec:
+                options.append('--with-gasnet=%s/src' % spec['gasnet'].prefix)
```
This could be `options.append('--with-gasnet=%s' % spec['gasnet'].prefix.src)`

bonachea replied:
Fixed as suggested in 2f2f6b6

tldahlgren replied:
Looks like the `cae743d` undid that fix.  Do you want to restore the fix?

bonachea replied:
oops - no I accidentally just fixed only one of two instances of the same pattern.

Second instance fixed in d40594b

bonachea committed:
upcxx: Respond to PR feedback

bonachea commented:
I believe I've addressed all the comments and requested changes, and this is ready for hopefully final review. @becker33 and @tldahlgren, please have another look.

We are trying to get this merged by Friday to meet a deadline for an impending ECP E4S release.

Thanks!

tldahlgren reviewed (changes requested):
LGTM *except* it looks like a fix was accidentally replaced in the upcxx package.  

Did you want to re-do the fix?

bonachea committed:
upcxx: Respond to PR feedback x 2

tldahlgren reviewed (approved):
LGTM
